initial commit

yiocardamone opened the pull request.

Commit:
initial commit

syols commented on authorithy.py:
```diff
@@ -0,0 +1,52 @@
+    if 'Log Out' in html:  # Если строка 'Log Out' есть в html, значит авторизация прошла успешно
+        print('Login OK!')
+    else:
+        print('Login Error!')
```
В конце файла с исходным кодом необходимо добавлять пустую строку. Тут и далее

yiocardamone replied:
+

syols commented on main.py:
```diff
@@ -0,0 +1,19 @@
+    return "Index Page"
+
+
+@app.route("/weather")  # здесь один эндпоинт он один общий можно добавлять, но они не меняются
```
Лучше  ручки вынести в отдельный фаил

yiocardamone replied:
+

syols replied:
Ничего не изменилось?

syols commented on main.py:
```diff
@@ -0,0 +1,19 @@
+
+app = Flask(__name__)
+
+api_key = "f5b39ac7a82e4dbd8a5120339231312"  # Paste Your API ID Here
```
Приватная информация, в идеале, должна быть спрятана
Лучше перенести это в конфиг, а в самом конфиге залей "Paste Your API ID Here".

yiocardamone replied:
+

syols commented on main.py:
```diff
@@ -0,0 +1,19 @@
+app = Flask(__name__)
+
+api_key = "f5b39ac7a82e4dbd8a5120339231312"  # Paste Your API ID Here
+final_URL = "http://api.weatherapi.com/v1/current.json?key=" + api_key + "&q={}&aqi={}"
```
Эта информация не должна быть тут.

yiocardamone replied:
+

syols commented on main.py:
```diff
@@ -0,0 +1,19 @@
+app = Flask(__name__)
+
+api_key = "f5b39ac7a82e4dbd8a5120339231312"  # Paste Your API ID Here
+final_URL = "http://api.weatherapi.com/v1/current.json?key=" + api_key + "&q={}&aqi={}"
```
Почитай про формат строк через f""

yiocardamone replied:
+

syols commented on main.py:
```diff
@@ -0,0 +1,19 @@
+def getting():
+    # handlers внутри сервера на flask
+    result = TestClassModel('Moscow').request_conversion()
+    return str(result.temperature) + " " + str(result.lat) + " " + str(result.lon)
```
Поправь формат строк

yiocardamone replied:
+

syols commented on model/geo_temperature.py:
```diff
@@ -0,0 +1,8 @@
+from dataclasses import dataclass
```
Лучше переименовать в models

yiocardamone replied:
+

syols commented on model/geo_temperature.py:
```diff
@@ -0,0 +1,8 @@
+
+@dataclass
+class GeoTemperature:
+    temperature: float
```
Значения по умолчанию?

yiocardamone replied:
+

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,31 @@
+    IWeatherRequester):  # инстансы этих прослоек хранятся в сервере, это к вопросу о связи сервера и прослойки
+    # сама прослойка просто посылает реквесты
+
+    API_KEY = "f5b39ac7a82e4dbd8a5120339231312"  # Paste Your API ID Here
```
Дубли в коде

yiocardamone replied:
+

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,31 @@
+    # сама прослойка просто посылает реквесты
+
+    API_KEY = "f5b39ac7a82e4dbd8a5120339231312"  # Paste Your API ID Here
+    FINAL_URL = "http://api.weatherapi.com/v1/current.json?key=" + API_KEY + "&q={}&aqi={}"
```
Это приватная информация?

yiocardamone replied:
+

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,31 @@
+
+    def request_conversion(self):
+        result = requests.get(self.FINAL_URL.format(self.city, self.API_KEY))
+        data = result.json()
```
Тут может быть исключение, надо обработать

yiocardamone replied:
+

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,31 @@
+    # сама прослойка просто посылает реквесты
+
+    API_KEY = "f5b39ac7a82e4dbd8a5120339231312"  # Paste Your API ID Here
+    FINAL_URL = "http://api.weatherapi.com/v1/current.json?key=" + API_KEY + "&q={}&aqi={}"
```
URL так не строят

syols replied:
Посмотри: `urllib.parse.urlparse`

yiocardamone replied:
+

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,31 @@
+        result = requests.get(self.FINAL_URL.format(self.city, self.API_KEY))
+        data = result.json()
+
+        temperature = data['current']['temp_c']
```
А что будет, если с сервера этот ключ не придет? см `get`

yiocardamone replied:
-

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,31 @@
+        coordinate_lon = data['location']['lon']
+        coordinate_lat = data['location']['lat']
+
+        return str(temperature) + " " + str(coordinate_lat) + " " + str(coordinate_lon)
```
Форматирование

yiocardamone replied:
+

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,31 @@
+        pass
+
+
+if __name__ == '__main__':
```
Зачем это тут?

yiocardamone replied:
+

syols commented on model/test_class_model.py:
```diff
@@ -0,0 +1,20 @@
+from abc import ABC
```
Перенеси это в папку tests

yiocardamone replied:
+

syols commented on model/test_class_model.py:
```diff
@@ -0,0 +1,20 @@
+from abc import ABC
```
А где сами тесты?

syols commented on model/test_class_model.py:
```diff
@@ -0,0 +1,20 @@
+from model.weather_requester import IWeatherRequester
+
+
+class TestClassModel(IWeatherRequester, ABC):
```
ClassModel?
Мб OpenWeatherRequesterMock?

yiocardamone replied:
+

syols commented on model/weather_requester.py:
```diff
@@ -0,0 +1,13 @@
+class IWeatherRequester(ABC):
+    @abstractmethod
+    def request_conversion(self):
+        pass
```
замени на `raise NotImplementedError`

yiocardamone replied:
+

syols commented on model/weather_requester.py:
```diff
@@ -0,0 +1,13 @@
+
+class IWeatherRequester(ABC):
+    @abstractmethod
+    def request_conversion(self):
```
Добавь тут и во всем проекте [type hints](https://docs.python.org/3/library/typing.html)

yiocardamone replied:
+

syols replied:
Нет, это не исправлено. Не хватает возвращаемых значений. 
Попробуй проверить тайпинг через mypy, подсветит все проблемы

syols commented on model/weather_requester.py:
```diff
@@ -0,0 +1,13 @@
+
+    @abstractmethod
+    def response_conversion(self):
+        pass
```
Почитай про утилиты isort, mypy и black. Проверь свой код

yiocardamone replied:
+

syols replied:
Это не исправлено (mypy, black)

syols commented on main.py:
```diff
@@ -0,0 +1,19 @@
+from flask import Flask
+from model.test_class_model import *
```
не импортируй со *

yiocardamone replied:
+

syols commented on authorithy.py:
```diff
@@ -0,0 +1,52 @@
+import time
```
Для чего этот фаил?

syols commented on main.py:
```diff
@@ -0,0 +1,19 @@
+from flask import Flask
```
Тут используются зависимости, почитай про pip install и [managing-dependencies](https://www.jetbrains.com/help/pycharm/managing-dependencies.html)

В идеале попробовать поработать с poetry

syols replied:
pip freeze

syols commented on model/test_class_model.py:
```diff
@@ -0,0 +1,20 @@
+        self.city = city
+
+    def response_conversion(self):
+        tom = GeoTemperature(self.temperature, self.lat, self.lon)
```
мб назвать иначе? tom?

yiocardamone replied:
+

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,31 @@
+    FINAL_URL = "http://api.weatherapi.com/v1/current.json?key=" + API_KEY + "&q={}&aqi={}"
+
+    def __init__(self, city):
+        self.city = city
```
Все переменные публичные?

yiocardamone replied:
+

syols commented on main.py:
```diff
@@ -0,0 +1,19 @@
+
+
+@app.route("/")
+def index():
```
А зачем эта ручка?

yiocardamone replied:
+

syols commented on model/test_class_model.py:
```diff
@@ -0,0 +1,20 @@
+
+    def response_conversion(self):
+        tom = GeoTemperature(self.temperature, self.lat, self.lon)
+        return tom
```
можно же сразу вернуть `GeoTemperature`

yiocardamone replied:
+

syols commented on main.py:
```diff
@@ -0,0 +1,19 @@
+def getting():
+    # handlers внутри сервера на flask
+    result = TestClassModel('Moscow').request_conversion()
+    return str(result.temperature) + " " + str(result.lat) + " " + str(result.lon)
```
Тут лучше вернуть json

yiocardamone replied:
+

syols commented on main.py:
```diff
@@ -0,0 +1,19 @@
+    return "Index Page"
+
+
+@app.route("/weather")  # здесь один эндпоинт он один общий можно добавлять, но они не меняются
```
Забыли HTTP метод

yiocardamone replied:
+

syols reviewed (commented):
Ревью кода, остальную логику (там много вопросов) лучше посмотреть после текущих правок

Commit:
Create a separate file endpoints.py.
Write private information to the config.py file and add it to the .gitignore exceptions.
Return the geo_temperature object in json format.

syols commented on .gitignore:
```diff
@@ -2,4 +2,7 @@
+.idea/
+
+# Private api_key
+config.py
```
Конфиг лучше в json/yaml и хорошо бы его залить, просто без ключей

yiocardamone replied:
+

syols commented on endpoints.py:
```diff
@@ -0,0 +1,13 @@
+
+import json
+
+app = Flask(__name__)
```
А зачем тут держать app?

yiocardamone replied:
+

syols commented on model/simplified_inheritance.py:
```diff
@@ -1,27 +1,28 @@
-class SimplifiedInheritance(
-    IWeatherRequester):  # инстансы этих прослоек хранятся в сервере, это к вопросу о связи сервера и прослойки
-    # сама прослойка просто посылает реквесты
+class SimplifiedInheritance(IWeatherRequester, ABC):  # instance этих прослоек хранятся в сервере, это к вопросу о
```
ABC тут точно нужно?

yiocardamone replied:
+

Commit:
Refine the "Simplified Inheritance" class to access the available APIs from the list.

syols commented on config.json:
Добавь сам json плиз

syols commented on authorithy.py:
```diff
@@ -0,0 +1,56 @@
+    "utf8": "✓",
+    "authenticity_token": authenticity_token,
+    "user[email]": "desoud.ck@gmail.com",  # Email
+    "user[password]": "18283848Qw",  # Пароль
```
:) как думаешь, пароль рабочий?

syols replied:
Да, рабочий. Интересно, а где еще эта связка email/password используется?

syols commented on authorithy.py:
```diff
@@ -0,0 +1,56 @@
+# Получение токена
+response = session.post(url, data)
+
+time.sleep(2)  # Пауза 2 сек :)
```
Это очень плохо. Даже если в тестах есть таймауты, значит что-то не так

syols commented on authorithy.py:
```diff
@@ -0,0 +1,56 @@
+time.sleep(2)  # Пауза 2 сек :)
+html = response.text  # Авторизуемся. В html будет наш ответ после авторизации
+
+if __name__ == "__main__":
```
А зачем тут __main__?

syols commented on main.py:
```diff
@@ -0,0 +1,12 @@
+
+
+# регистрируем обработчик
+app.add_url_rule("/weather", "get_weather", get_weather, methods=["GET"])
```
Ручку так и не вынесли

syols replied:
routes.get_weather - поменять в views.get_weather
app.add_url_rule("/weather", "get_weather", get_weather, methods=["GET"]) - в routes

syols commented on tests/open_weather_requester_mock.py:
```diff
@@ -0,0 +1,19 @@
+from model.weather_requester import IWeatherRequester
+
+
+class OpenWeatherRequesterMock(IWeatherRequester, ABC):
```
А где сами тесты?

syols replied:
ABC?

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,82 @@
+
+
+def find_values_by_key(data: dict, key: str):
+    def find_values(obj: dict):
```
Круто, что ты попробовал сделать свой десереализатор, но лучше взять pydantic/marshmallow.

syols replied:
сама логика проверки через isinstance спорная + рекурсия. Но хорошо, что попробовал

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,82 @@
+
+class SimplifiedInheritance(IWeatherRequester):
+    current_directory = os.getcwd()
+    path = os.path.join(current_directory, "config.json")
```
1. Разделение ролей: конфиг лучше передавать в параметрах и инициализировать единожды
2. Десереализацию конфига лучше сделать через pydantic/marshmallow и т.д

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,82 @@
+
+                try:
+                    result = requests.get(url.format(authenticity_key, self.coordinate))
+                    result.raise_for_status()  # Проверяем наличие ошибки в ответе.
```
Круто

syols replied:
но, эта конструкция должна быть где-то обработана, тк летит исключение (например в get_weather в routes.py)

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,82 @@
+
+                except requests.exceptions.HTTPError as err:
+                    if result.status_code == 400:
+                        print(f"URL-адрес запроса API недействителен: {err}")
```
Нет, лучше создать набор ошибок и raise'ить их тут.
В самом обработчике ручки /weather/ вылавливать ошибку и возвращать уже наш HTTP-код ошибки в ответе

syols replied:
далее тоже самое

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,82 @@
+                    print(f"Произошла ошибка обработки запроса: {err}")
+
+                data = result.json()
+                support_params = subjects["support_params"]
```
Опять же - десереализация

syols commented on routes.py:
```diff
@@ -0,0 +1,9 @@
+def get_weather():  # здесь один endpoint, он один - общий можно добавлять, но они не меняются.
+    # handlers внутри сервера на flask.
+    result = SimplifiedInheritance("55.75, 37.62").request_conversion()
+    return json.dumps(result.__dict__)
```
смотри flask.jsonify

syols commented on tests/open_weather_requester_mock.py:
```diff
@@ -0,0 +1,19 @@
+    lon: float = 37.62
+
+    def __init__(self, city: str):
+        self.city = city
```
публичные параметры, уже вроде правили

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,82 @@
+    path = os.path.join(current_directory, "config.json")
+
+    def __init__(self, coordinate: str):
+        self.coordinate = coordinate
```
Публичные параметры

syols commented on authorithy.py:
```diff
@@ -0,0 +1,56 @@
+    if (
+        "Log Out" in html
+    ):  # Если строка 'Log Out' есть в html, значит авторизация прошла успешно
+        print("Login OK!")
```
Предлагаю убрать принты, для этих целей используй логирование (к примеру, https://flask.palletsprojects.com/en/1.1.x/logging/)

syols replied:
app.logger и тд

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,82 @@
+        if isinstance(obj, dict):
+            for k, v in obj.items():
+                if k == key:
+                    if isinstance(v, list):
```
Посмотри что такое тернарный условный оператор

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,82 @@
+
+
+class SimplifiedInheritance(IWeatherRequester):
+    current_directory = os.getcwd()
```
Плохая практика, почитай про абсолютные/относительные пути

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,82 @@
+                location = list(support_params.keys())[0]  # object location
+                current = list(support_params.keys())[1]  # object temperature
+
+                lat = support_params[location][0]  # the name of the "location" field
```
так делать плохо, у тебя там должен быть tuple. 
lat/lon это пара значений, можно так:`lat, lon = support_params[location]`

Commit:
test

syols commented on config.json:
```diff
@@ -0,0 +1,24 @@
+    {
+      "url": "https://api.tomorrow.io/v4/weather/realtime?apikey={}&location={}",
+      "query_params": {
+        "apikey": "Sb5mxppt7iHIRRTMwXmV3SzVrNXT9AO9"
```
Убери ключи плз

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,71 @@
+
+def find_values_by_key(data: dict, key: str) -> Any:
+    def find_values(obj: dict) -> Any:
+        return next((v[0] for k, v in obj.items() if k == key and isinstance(v, list)), None) if isinstance(obj, dict) \
```
Нет десериализации, см комменты

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,71 @@
+
+
+class SimplifiedInheritance(IWeatherRequester):
+    path = "./config.json"
```
Публичный снова параметр

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,71 @@
+        self.coordinate = coordinate
+
+    def request_conversion(self):
+        with open(self.path) as f:
```
Каждый раз грузишь конфиг?

Commit:
Add serialization of the config and server response, make variables private, move the handler and endpoint.

syols commented on model/serialize_tomorrow.py:
```diff
@@ -0,0 +1,90 @@
+                 precipitationProbability, pressureSurfaceLevel, rainIntensity, sleetIntensity, snowIntensity,
+                 temperature, temperatureApparent, uvHealthConcern, uvIndex, visibility, weatherCode,
+                 windDirection, windGust, windSpeed):
+        self.cloudBase = cloudBase
```
Дубли со Schema, хорошо бы поправить

Commit:
Pass the configuration once in the parameters.

syols commented on model/simplified_inheritance.py:
```diff
@@ -10,52 +9,44 @@
-
-                result = schema.load(data)
-                return GeoTemperature(result[0], result[1], result[2])
+        for line in self.config.items():
```
можно сделать: for k, v in self.config.items():

В любом случае, конфиг надо десереализовать в класс и не забирать данные из дикта

syols replied:
`def __init__(self, coordinate: str, config: dict): ` выше config определен как dict

syols replied:
for line in self.config.items():
url = line[1]["url"]
key = line[1]["key"]

тоже dict

syols replied:
см TypedDict

Commit:
Serialization of the config. Exception handling.

syols commented on model/status_code.py:
```diff
@@ -0,0 +1,11 @@
+
+
+class StatusCode(Enum):
+    RESPONSE_CODE_BAD_REQUEST = 400
```
https://docs.python.org/3/library/http.html можно не изобретать

syols commented on serialize_config.py:
```diff
@@ -1,9 +1,52 @@
+    forecast: Forecast
+
+
+"""schema.py"""
```
это для примера было - можно удалить

syols commented on views.py:
```diff
@@ -1,19 +1,28 @@
+            data = json.load(f)
+        except ValueError:
+            logging.warning("Incorrect data in before_request")
+            raise
```
Неправильно рэйзить ошибку тут, надо либо ее тут перехватывать и только писать в лог - log.error, либо возвращать статус код в ручке (отличный от 500)

syols commented on views.py:
```diff
@@ -1,19 +1,28 @@
 
 def get_weather():  # здесь один endpoint, он один - общий можно добавлять, но они не меняются.
     # handlers внутри сервера на flask.
-    result = SimplifiedInheritance("55.75, 37.62", g.config).request_conversion()
```
Хардкод нужно убрать - в ручке должны быть эти query-параметры

syols commented on model/simplified_inheritance.py:
```diff
@@ -1,43 +1,75 @@
+        # Перехватываем HTTPError и создаем объект CustomHTTPError с пользовательским сообщением
+        if response.status_code == StatusCode.RESPONSE_CODE_BAD_REQUEST.value:
+            logging.warning(f"The API request URL is invalid. Status code {status_code}")
+            raise CustomException("The API request URL is invalid",
```
CustomException - не оч удачное название

syols commented on exceptions/weather_exception.py:
```diff
@@ -0,0 +1,14 @@
+class CustomException(Exception):
+    def __init__(self, message, status_code):
+        super().__init__(message)
+        self.message = message
```
Публичные параметры, ну сколько их уже? Ответь на этот вопрос, если читаешь ревью!

syols commented on views.py:
```diff
@@ -0,0 +1,28 @@
+
+def get_weather():  # здесь один endpoint, он один - общий можно добавлять, но они не меняются.
+    # handlers внутри сервера на flask.
+    result = SimplifiedInheritance("55.75, 37.62", g.result).request_conversion()
```
Хардкод

syols replied:
Обработка ошибок тут

syols commented on model/simplified_inheritance.py:
```diff
@@ -0,0 +1,84 @@
+    return data, continue_flag
+
+
+class SimplifiedInheritance(IWeatherRequester):
```
WeatherApiRequester

syols replied:
```
IWeatherRequester
    WeatherAPIRequester
    TomorrowIORequester

```

syols commented on serialize_config.py:
```diff
@@ -0,0 +1,52 @@
+import dataclasses
```
переименовать в config.py

syols commented on config.json:
```diff
@@ -0,0 +1,16 @@
+{
+  "arr": [
```
?

syols replied:
Структура:
```
{
   "api.weatherapi.com":
        {"key": "__INSERT_KEY_HERE__"},
   "api.tomorrow.io":
        {"key": "__INSERT_KEY_HERE__"}
}

```

syols commented on views.py:
```diff
@@ -0,0 +1,28 @@
+def before_request():
+    _path = "./config.json"
+
+    with open(_path, 'r') as f:
```
перенести в config.py

syols commented on views.py:
```diff
@@ -0,0 +1,28 @@
+from serialize_config import WeatherSchema
+
+
+def before_request():
```
тут только создание инстанса класса Config и передача в ручку

syols commented on serialize_config.py:
```diff
@@ -0,0 +1,52 @@
+
+"""models/weather.py"""
+
+@dataclasses.dataclass
```
Тут что-то не то

syols commented on model/weather_requester.py:
```diff
@@ -0,0 +1,11 @@
+from abc import ABC, abstractmethod
+
+
+class IWeatherRequester(ABC):
```
IWeatherRequesterBase

Commit:
Serialization of the config. Exception handling.

syols commented on errors/external_service_base_exception.py:
```diff
@@ -0,0 +1,46 @@
+        return self._message
+
+
+class ExternalServiceClientException(ExternalServiceBaseException):
```
сlass ExternalServiceClientException(BaseException):
    pass

syols replied:
зачем дубли

Commit:
- Fixed issues in the code.
- Removed hard coding.
- Added tests.
- Added mypy, black, and isort tools.
- Created the mypy.ini file to configure the type tool and exclude third-party libraries from type checking.

Commit:
- Duplicate entries were removed from the deserializer schema and exception classes.
- Errors were addressed using mypy . --disallow-untyped-defs.
